In [1]:
import re
import pdfplumber
import spacy
import logging
from spellchecker import SpellChecker
from fuzzywuzzy import fuzz

# Path to the PDF file
pdf_path = "C:/Users/PC/Documents/FLATIRON/PHASE5/Kenya-Constitution-Chatbot/Kenya Constitution.pdf"

with pdfplumber.open(pdf_path) as pdf:
    for page in pdf.pages:
        text = page.extract_text()

In [2]:
# Path to your PDF file
pdf_path = "C:/Users/PC/Documents/FLATIRON/PHASE5/Kenya-Constitution-Chatbot/Kenya Constitution.pdf"

# Function to extract text from specific pages
def extract_specific_pages(cleaned_constitution, start_page, end_page):
    with pdfplumber.open(cleaned_constitution) as pdf:
        extracted_text = ""
        for page_num in range(start_page, end_page):
            page = pdf.pages[page_num]
            extracted_text += page.extract_text() + "\n"
        return extracted_text

### Chapter 17
### General Provisions

In [3]:
# extracting the whole of chapter 4 which is located in page 18 to 41
chapter_17 = extract_specific_pages(pdf_path, 156,164 )
chapter_17

'Constitution of Kenya, 2010 157\n(8) A Bill under this Article is passed by Parliament if supported\nby a majority of the members of each House.\n(9) If Parliament passes the Bill, it shall be submitted to the\nPresident for assent in accordance with Articles 256 (4) and (5).\n(10) If either House of Parliament fails to pass the Bill, or the Bill\nrelates to a matter specified in 255 (1), the proposed amendment shall\nbe submitted to the people in a referendum.\n(11) Article 255 (2) applies, with any necessary modifications, to\na referendum under clause (10).\nCHAPTER SEVENTEEN—GENERAL PROVISIONS\nEnforcement of this Constitution.\n258. (1) Every person has the right to institute court proceedings,\nclaiming that this Constitution has been contravened, or is threatened\nwith contravention.\n(2) In addition to a person acting in their own interest, court\nproceedings under clause (1) may be instituted by—\n(a) a person acting on behalf of another person who cannot act in\ntheir own na

### The chapter 17 starts at page 157, which is shared by the end of chapter 16 and ends at page 157 and still the page is split into two thats is one part of the page is the start of chapter 18, we have to split this inorder to have just the info of chapter 17.

In [4]:
# Extract only the content of CHAPTER SEVENTEEN without including CHAPTER EIGHTEEN
if "CHAPTER SEVENTEEN" in chapter_17:
    if "CHAPTER EIGHTEEN" in chapter_17:
        chapter_17_trimmed = chapter_17.split("CHAPTER SEVENTEEN", 1)[1].split("CHAPTER EIGHTEEN", 1)[0].strip()
    else:
        chapter_17_trimmed = chapter_17.split("CHAPTER SEVENTEEN", 1)[1].strip()

    print(chapter_17_trimmed)

—GENERAL PROVISIONS
Enforcement of this Constitution.
258. (1) Every person has the right to institute court proceedings,
claiming that this Constitution has been contravened, or is threatened
with contravention.
(2) In addition to a person acting in their own interest, court
proceedings under clause (1) may be instituted by—
(a) a person acting on behalf of another person who cannot act in
their own name;
(b) a person acting as a member of, or in the interest of, a group
or class of persons;
(c) a person acting in the public interest; or
(d) an association acting in the interest of one or more of its
members.
Construing this Constitution.
259. (1) This Constitution shall be interpreted in a manner that—
(a) promotes its purposes, values and principles;
(b) advances the rule of law, and the human rights and
fundamental freedoms in the Bill of Rights;
(c) permits the development of the law; and
(d) contributes to good governance.
(2) If there is a conflict between different language ver

In [5]:
# Load spaCy model
nlp = spacy.load("en_core_web_sm")

# Preprocess the user query using spaCy
def preprocess_query(query):
    # Parse the query with spaCy
    doc = nlp(query)
    # Normalize the query: lowercase, lemmatize, and remove stopwords
    tokens = [token.lemma_.lower() for token in doc if not token.is_stop and not token.is_punct]
    return " ".join(tokens)

# Example usage of the preprocessing function
user_query = "What is the supremacy of the constitution?"
processed_query = preprocess_query(user_query)
print(processed_query)

supremacy constitution


In [6]:
def split_chapter_17(chapter_text):
    # Split at key headings and strip extra whitespace
    sections = {
        "Enforcement of this Constitution": [],
        "Construing this Constitution": [],
        "Interpretation": [] 
    }

    # Split by new lines to process line by line
    lines = chapter_text.splitlines()

    current_section = None

    for line in lines:
        stripped_line = line.strip()

        if stripped_line.startswith("Enforcement of this Constitution"):
            current_section = "Enforcement of this Constitution"
        elif stripped_line.startswith("Construing this Constitution"):
            current_section = "Construing this Constitution"
        elif stripped_line.startswith("Interpretation"):
            current_section = "Interpretation"
        
        # Append line to the current section if it's set
        if current_section:
            sections[current_section].append(stripped_line)

    # Join each section into a single string
    for key in sections:
        sections[key] = "\n".join(sections[key])

    return sections

# Split the chapter into sections
chapter_17_sections = split_chapter_17(chapter_17_trimmed)

print("Enforcement of this Constitution Section:\n", chapter_17_sections['Enforcement of this Constitution'])
print("\nConstruing this Constitution Section:\n", chapter_17_sections['Construing this Constitution'])
print("\nInterpretation Section:\n", chapter_17_sections['Interpretation'])

Enforcement of this Constitution Section:
 Enforcement of this Constitution.
258. (1) Every person has the right to institute court proceedings,
claiming that this Constitution has been contravened, or is threatened
with contravention.
(2) In addition to a person acting in their own interest, court
proceedings under clause (1) may be instituted by—
(a) a person acting on behalf of another person who cannot act in
their own name;
(b) a person acting as a member of, or in the interest of, a group
or class of persons;
(c) a person acting in the public interest; or
(d) an association acting in the interest of one or more of its
members.

Construing this Constitution Section:
 Construing this Constitution.
259. (1) This Constitution shall be interpreted in a manner that—
(a) promotes its purposes, values and principles;
(b) advances the rule of law, and the human rights and
fundamental freedoms in the Bill of Rights;
(c) permits the development of the law; and
(d) contributes to good govern

In [7]:
chapter_17_sections

{'Enforcement of this Constitution': 'Enforcement of this Constitution.\n258. (1) Every person has the right to institute court proceedings,\nclaiming that this Constitution has been contravened, or is threatened\nwith contravention.\n(2) In addition to a person acting in their own interest, court\nproceedings under clause (1) may be instituted by—\n(a) a person acting on behalf of another person who cannot act in\ntheir own name;\n(b) a person acting as a member of, or in the interest of, a group\nor class of persons;\n(c) a person acting in the public interest; or\n(d) an association acting in the interest of one or more of its\nmembers.',
 'Construing this Constitution': 'Construing this Constitution.\n259. (1) This Constitution shall be interpreted in a manner that—\n(a) promotes its purposes, values and principles;\n(b) advances the rule of law, and the human rights and\nfundamental freedoms in the Bill of Rights;\n(c) permits the development of the law; and\n(d) contributes to go

In [8]:
# Define the sections variable with Chapter 1
sections = chapter_17_sections  # Assuming chapter_1 contains the text extracted from the PDF

# Define the QA mapping based on key phrases and corresponding sections
qa_mapping = {
    "Enforcement of this Constitution": "Enforcement of this Constitution",
    "Construing this Constitution": "Construing this Constitution", 
    "Interpretation": "Interpretation"
}

# Update the Q&A system to use preprocessed queries
def preprocess_query(query):
    return query.lower()  # Ensure case-insensitive matching

# Fix the answer_question_nlp function
def answer_question_nlp(query):
    # Preprocess the query
    preprocessed_query = preprocess_query(query)

    # Search for a matching key in QA_mapping
    for key, value in qa_mapping.items():
        if key.lower() in preprocessed_query:
            # Return the relevant section using the shorter section key (e.g., "Rights")
            return sections[key]  

    return "Sorry, I couldn't find an answer to your question."

# Example query
user_query = "Enforcement of this Constitution"
answer = answer_question_nlp(user_query)
print(answer)

Enforcement of this Constitution.
258. (1) Every person has the right to institute court proceedings,
claiming that this Constitution has been contravened, or is threatened
with contravention.
(2) In addition to a person acting in their own interest, court
proceedings under clause (1) may be instituted by—
(a) a person acting on behalf of another person who cannot act in
their own name;
(b) a person acting as a member of, or in the interest of, a group
or class of persons;
(c) a person acting in the public interest; or
(d) an association acting in the interest of one or more of its
members.


In [9]:
user_query = "Construing this Constitution"
answer = answer_question_nlp(user_query)
print(answer)

Construing this Constitution.
259. (1) This Constitution shall be interpreted in a manner that—
(a) promotes its purposes, values and principles;
(b) advances the rule of law, and the human rights and
fundamental freedoms in the Bill of Rights;
(c) permits the development of the law; and
(d) contributes to good governance.
(2) If there is a conflict between different language versions of
this Constitution, the English language version prevails.
158 Constitution of Kenya, 2010
(3) Every provision of this Constitution shall be construed
according to the doctrine of interpretation that the law is always
speaking and, therefore, among other things—
(a) a function or power conferred by this Constitution on an office
may be performed or exercised as occasion requires, by the
person holding the office;
(b) any reference in this Constitution to a State or other public
office or officer, or a person holding such an office, includes a
reference to the person acting in or otherwise performing the

In [10]:
user_query = "Interpretation"
answer = answer_question_nlp(user_query)
print(answer)

Interpretation.
260. In this Constitution, unless the context requires otherwise—
“adult” means an individual who has attained the age of eighteen
years;
“affirmative action” includes any measure designed to overcome
or ameliorate an inequity or the systemic denial or infringement of a
right or fundamental freedom;
160 Constitution of Kenya, 2010
“child” means an individual who has not attained the age of
eighteen years;
“contravene” includes fail to comply with;
“county legislation” means a law made by a county government
or under authority conferred by a county Assembly;
“disability” includes any physical, sensory, mental, psychological
or other impairment, condition or illness that has, or is perceived by
significant sectors of the community to have, a substantial or long-term
effect on an individual’s ability to carry out ordinary day-to-day
activities;
“document” includes—
(a) any publication, or any matter written, expressed, or inscribed
on any substance by means of letters, fig

### Debugging code

In [11]:
# Define synonym mapping
synonyms = {
    "Enforcement of this Constitution": ["Constitution enforcement", "implementation of Constitution", "constitutional enforcement"],
    "Construing this Constitution": ["interpreting this Constitution", "constitution interpretation", "Constitution reading"],
    "Interpretation": ["construing", "understanding", "deciphering"]
}

# QA mapping
qa_mapping = {
    "Enforcement of this Constitution": ("Enforcement of this Constitution"),
    "Construing this Constitution": ("Construing this Constitution"),
    "Interpretation": ("Interpretation")
}

# Preprocess the query
def preprocess_query(query):
    doc = nlp(query)
    tokens = [token.lemma_.lower() for token in doc if not token.is_stop and not token.is_punct]
    return " ".join(tokens)

# Function to correct the spelling
def correct_spelling(processed_query):
    spell = SpellChecker()
    words = processed_query.split()
    # Find misspelled words
    misspelled_words = spell.unknown(words)


    corrected_words = []
    for word in words:
        # Correct the word if it's misspelled
        if word in misspelled_words:
            corrected_word = spell.correction(word) # Get the most likely correction
            corrected_words.append(corrected_word)
        else:
            corrected_words.append(word)

    # Reconstruct the initial sentence
    corrected_input = " ".join(corrected_words)

    return corrected_input

            

# Match with fuzzy matching and synonym support
def match_with_synonyms(query, qa_mapping, synonyms):
    processed_query = preprocess_query(query)
    print(f"Processed Query: {processed_query}")  # Debugging line

    # Correct spelling in the processed query
    corrected_query = correct_spelling(processed_query)
    print(f"Corrected Query: {corrected_query}")  # Debugging line


    for key, value in qa_mapping.items():
        print(f"Checking key: {key}, value: {value}")  # Debugging line
        for synonym in synonyms.get(key, [key]):
            print(f"Trying synonym: {synonym}")  # Debugging line
            if synonym in corrected_query:
                print(f"Match found with synonym: {synonym}")  # Debugging line
                return value  # Only return section value of the key
            
            # Check for fuzzy matching with the synonym
            if fuzz.ratio(synonym, corrected_query) > 70:
                print(f"Fuzzy match found with synonym: {synonym}")  # Debugging line
                return value  # Return the key for the section
    
    print("No match found")  # Debugging line if no match is found
    return None

# Answer function with synonym and fuzzy matching
def answer_question_nlp(query, sections, qa_mapping, synonyms):
    section_key = match_with_synonyms(query, qa_mapping, synonyms)
    
    if section_key:
        # Retrieve the relevant section from the specified chapter
        return sections.get(section_key, "Section not found.")
    
    return "Sorry, I couldn't find an answer to your question."

### Precise code

In [12]:
# Define synonym mapping
synonyms = {
    "Enforcement of this Constitution": ["Constitution enforcement", "implementation of Constitution", "constitutional enforcement"],
    "Construing this Constitution": ["interpreting this Constitution", "constitution interpretation", "Constitution reading"],
    "Interpretation": ["construing", "understanding", "deciphering"]
}

# QA mapping
qa_mapping = {
    "Enforcement of this Constitution": ("Enforcement of this Constitution"),
    "Construing this Constitution": ("Construing this Constitution"),
    "Interpretation": ("Interpretation")
}

# Preprocess the query
def preprocess_query(query):
    doc = nlp(query)
    tokens = [token.lemma_.lower() for token in doc if not token.is_stop and not token.is_punct]
    return " ".join(tokens)

# Function to correct the spelling
def correct_spelling(processed_query):
    spell = SpellChecker()
    words = processed_query.split()
    # Find misspelled words
    misspelled_words = spell.unknown(words)


    corrected_words = []
    for word in words:
        # Correct the word if it's misspelled
        if word in misspelled_words:
            corrected_word = spell.correction(word) # Get the most likely correction
            corrected_words.append(corrected_word)
        else:
            corrected_words.append(word)

    # Reconstruct the initial sentence
    corrected_input = " ".join(corrected_words)

    return corrected_input

            

# Match with fuzzy matching and synonym support
def match_with_synonyms(query, qa_mapping, synonyms):
    processed_query = preprocess_query(query)

    # Correct spelling in the processed query
    corrected_query = correct_spelling(processed_query)


    for key, value in qa_mapping.items():
        for synonym in synonyms.get(key, [key]):
            if synonym in corrected_query:
                return value  # Only return section value of the key
            
            # Check for fuzzy matching with the synonym
            if fuzz.ratio(synonym, corrected_query) > 70:
                print(f"Fuzzy match found with synonym: {synonym}")  # Debugging line
                return value  # Return the key for the section
    
    print("No match found")
    return None

# Answer function with synonym and fuzzy matching
def answer_question_nlp(query, sections, qa_mapping, synonyms):
    section_key = match_with_synonyms(query, qa_mapping, synonyms)
    
    if section_key:
        # Retrieve the relevant section from the specified chapter
        return sections.get(section_key, "Section not found.")
    
    return "Sorry, I couldn't find an answer to your question."

In [13]:
sections = chapter_17_sections

In [14]:
user_query = "what does the constitution say about implementation of Constitution?"
answer = answer_question_nlp(user_query, sections, qa_mapping, synonyms)
print(answer)

Fuzzy match found with synonym: implementation of Constitution
Enforcement of this Constitution.
258. (1) Every person has the right to institute court proceedings,
claiming that this Constitution has been contravened, or is threatened
with contravention.
(2) In addition to a person acting in their own interest, court
proceedings under clause (1) may be instituted by—
(a) a person acting on behalf of another person who cannot act in
their own name;
(b) a person acting as a member of, or in the interest of, a group
or class of persons;
(c) a person acting in the public interest; or
(d) an association acting in the interest of one or more of its
members.


In [15]:
user_query = "explain the meaning constitution interpretation?"
answer = answer_question_nlp(user_query, sections, qa_mapping, synonyms)
print(answer)

Construing this Constitution.
259. (1) This Constitution shall be interpreted in a manner that—
(a) promotes its purposes, values and principles;
(b) advances the rule of law, and the human rights and
fundamental freedoms in the Bill of Rights;
(c) permits the development of the law; and
(d) contributes to good governance.
(2) If there is a conflict between different language versions of
this Constitution, the English language version prevails.
158 Constitution of Kenya, 2010
(3) Every provision of this Constitution shall be construed
according to the doctrine of interpretation that the law is always
speaking and, therefore, among other things—
(a) a function or power conferred by this Constitution on an office
may be performed or exercised as occasion requires, by the
person holding the office;
(b) any reference in this Constitution to a State or other public
office or officer, or a person holding such an office, includes a
reference to the person acting in or otherwise performing the

In [16]:
user_query = "briefly explain deciphering?"
answer = answer_question_nlp(user_query, sections, qa_mapping, synonyms)
print(answer)

Interpretation.
260. In this Constitution, unless the context requires otherwise—
“adult” means an individual who has attained the age of eighteen
years;
“affirmative action” includes any measure designed to overcome
or ameliorate an inequity or the systemic denial or infringement of a
right or fundamental freedom;
160 Constitution of Kenya, 2010
“child” means an individual who has not attained the age of
eighteen years;
“contravene” includes fail to comply with;
“county legislation” means a law made by a county government
or under authority conferred by a county Assembly;
“disability” includes any physical, sensory, mental, psychological
or other impairment, condition or illness that has, or is perceived by
significant sectors of the community to have, a substantial or long-term
effect on an individual’s ability to carry out ordinary day-to-day
activities;
“document” includes—
(a) any publication, or any matter written, expressed, or inscribed
on any substance by means of letters, fig

In [17]:
chapter_17_sections.keys()

dict_keys(['Enforcement of this Constitution', 'Construing this Constitution', 'Interpretation'])

In [18]:
# Load the spacy model
nlp = spacy.load("en_core_web_sm")

# Configure logging
logging.basicConfig(
    level=logging.DEBUG,  # Set the logging level
    format='%(asctime)s - %(levelname)s - %(message)s',  # Log message format
)

# Define synonym mapping
synonyms = {
    "Enforcement of this Constitution": ["Constitution enforcement", "implementation of Constitution", "constitutional enforcement"],
    "Construing this Constitution": ["interpreting this Constitution", "constitution interpretation", "Constitution reading"],
    "Interpretation": ["construing", "understanding", "deciphering"]
}

# QA mapping
qa_mapping = {
    "Enforcement of this Constitution": ("Enforcement of this Constitution"),
    "Construing this Constitution": ("Construing this Constitution"),
    "Interpretation": ("Interpretation")
}


# Preprocess the query
def preprocess_query(query):
    doc = nlp(query)
    tokens = [token.lemma_.lower() for token in doc if not token.is_stop and not token.is_punct]
    return " ".join(tokens)

# Function to correct the spelling
def correct_spelling(processed_query):
    spell = SpellChecker()
    words = processed_query.split()
    # Find misspelled words
    misspelled_words = spell.unknown(words)

    corrected_words = []
    for word in words:
        # Correct the word if it's misspelled
        if word in misspelled_words:
            corrected_word = spell.correction(word)  # Get the most likely correction
            corrected_words.append(corrected_word)
        else:
            corrected_words.append(word)

    # Reconstruct the initial sentence
    corrected_input = " ".join(corrected_words)
    return corrected_input

# Match the words with synonym support
def match_with_synonyms(query, qa_mapping, synonyms, threshold=70):
    processed_query = preprocess_query(query)
    logging.debug(f"Processed Query: {processed_query}")  # Debugging line

    # Correct spelling in the processed query
    corrected_query = correct_spelling(processed_query)
    logging.debug(f"Corrected Query: {corrected_query}")  # Debugging line

    for key, value in qa_mapping.items():
        logging.debug(f"Checking key: {key}, value: {value}")  # Debugging line
        for synonym in synonyms.get(key, [key]):
            logging.debug(f"Trying synonym: {synonym}")  # Debugging line
            if synonym in corrected_query:
                logging.debug(f"Match found with synonym: {synonym}")  # Debugging line
                return value  # Only return section value of the key
            
            # Check for fuzzy matching with the synonym using the threshold parameter
            if fuzz.ratio(synonym, corrected_query) > threshold:
                logging.debug(f"Fuzzy match found with synonym: {synonym}")  # Debugging line
                return value  # Return the key for the section
    
    logging.debug("No match found")  # Debugging line if no match is found
    return None

# Answer function with synonym and fuzzy matching
def answer_question_nlp(query, sections, qa_mapping, synonyms, threshold=70):
    section_key = match_with_synonyms(query, qa_mapping, synonyms, threshold)
    
    if section_key:
        # Retrieve the relevant section from the specified chapter
        return sections.get(section_key, "Section not found.")
    
    logging.error("No answer found for the question.")  # Log an error if no answer is found
    return "Sorry, I couldn't find an answer to your question."

# Sample usage
if __name__ == "__main__":
    user_query = "What is Constitution reading?"
    threshold_value = 75  # Adjust this value as needed
    answer = answer_question_nlp(user_query, sections, qa_mapping, synonyms, threshold=threshold_value)
    print(answer)

2024-10-31 01:53:59,751 - DEBUG - Processed Query: constitution reading
2024-10-31 01:54:00,133 - DEBUG - Corrected Query: constitution reading
2024-10-31 01:54:00,133 - DEBUG - Checking key: Enforcement of this Constitution, value: Enforcement of this Constitution
2024-10-31 01:54:00,142 - DEBUG - Trying synonym: Constitution enforcement
2024-10-31 01:54:00,145 - DEBUG - Trying synonym: implementation of Constitution
2024-10-31 01:54:00,148 - DEBUG - Trying synonym: constitutional enforcement
2024-10-31 01:54:00,150 - DEBUG - Checking key: Construing this Constitution, value: Construing this Constitution
2024-10-31 01:54:00,151 - DEBUG - Trying synonym: interpreting this Constitution
2024-10-31 01:54:00,153 - DEBUG - Trying synonym: constitution interpretation
2024-10-31 01:54:00,155 - DEBUG - Fuzzy match found with synonym: constitution interpretation


Construing this Constitution.
259. (1) This Constitution shall be interpreted in a manner that—
(a) promotes its purposes, values and principles;
(b) advances the rule of law, and the human rights and
fundamental freedoms in the Bill of Rights;
(c) permits the development of the law; and
(d) contributes to good governance.
(2) If there is a conflict between different language versions of
this Constitution, the English language version prevails.
158 Constitution of Kenya, 2010
(3) Every provision of this Constitution shall be construed
according to the doctrine of interpretation that the law is always
speaking and, therefore, among other things—
(a) a function or power conferred by this Constitution on an office
may be performed or exercised as occasion requires, by the
person holding the office;
(b) any reference in this Constitution to a State or other public
office or officer, or a person holding such an office, includes a
reference to the person acting in or otherwise performing the

In [19]:
user_query = "What is Constitution reading?"
threshold_value = 75  # Adjust this value as needed
answer = answer_question_nlp(user_query, sections, qa_mapping, synonyms, threshold=threshold_value)
print(answer)

2024-10-31 01:54:00,198 - DEBUG - Processed Query: constitution reading
2024-10-31 01:54:00,495 - DEBUG - Corrected Query: constitution reading
2024-10-31 01:54:00,495 - DEBUG - Checking key: Enforcement of this Constitution, value: Enforcement of this Constitution
2024-10-31 01:54:00,495 - DEBUG - Trying synonym: Constitution enforcement
2024-10-31 01:54:00,506 - DEBUG - Trying synonym: implementation of Constitution
2024-10-31 01:54:00,507 - DEBUG - Trying synonym: constitutional enforcement
2024-10-31 01:54:00,510 - DEBUG - Checking key: Construing this Constitution, value: Construing this Constitution
2024-10-31 01:54:00,511 - DEBUG - Trying synonym: interpreting this Constitution
2024-10-31 01:54:00,514 - DEBUG - Trying synonym: constitution interpretation
2024-10-31 01:54:00,514 - DEBUG - Fuzzy match found with synonym: constitution interpretation


Construing this Constitution.
259. (1) This Constitution shall be interpreted in a manner that—
(a) promotes its purposes, values and principles;
(b) advances the rule of law, and the human rights and
fundamental freedoms in the Bill of Rights;
(c) permits the development of the law; and
(d) contributes to good governance.
(2) If there is a conflict between different language versions of
this Constitution, the English language version prevails.
158 Constitution of Kenya, 2010
(3) Every provision of this Constitution shall be construed
according to the doctrine of interpretation that the law is always
speaking and, therefore, among other things—
(a) a function or power conferred by this Constitution on an office
may be performed or exercised as occasion requires, by the
person holding the office;
(b) any reference in this Constitution to a State or other public
office or officer, or a person holding such an office, includes a
reference to the person acting in or otherwise performing the

In [20]:
user_query = "What constitutional enforcement?"
threshold_value = 75  # Adjust this value as needed
answer = answer_question_nlp(user_query, sections, qa_mapping, synonyms, threshold=threshold_value)
print(answer)

2024-10-31 01:54:00,546 - DEBUG - Processed Query: constitutional enforcement
2024-10-31 01:54:00,854 - DEBUG - Corrected Query: constitutional enforcement
2024-10-31 01:54:00,855 - DEBUG - Checking key: Enforcement of this Constitution, value: Enforcement of this Constitution
2024-10-31 01:54:00,855 - DEBUG - Trying synonym: Constitution enforcement
2024-10-31 01:54:00,861 - DEBUG - Fuzzy match found with synonym: Constitution enforcement


Enforcement of this Constitution.
258. (1) Every person has the right to institute court proceedings,
claiming that this Constitution has been contravened, or is threatened
with contravention.
(2) In addition to a person acting in their own interest, court
proceedings under clause (1) may be instituted by—
(a) a person acting on behalf of another person who cannot act in
their own name;
(b) a person acting as a member of, or in the interest of, a group
or class of persons;
(c) a person acting in the public interest; or
(d) an association acting in the interest of one or more of its
members.
